sup3r/preprocessing/data_handling/dual_data_handling.py

```diff
@@ -0,0 +1,614 @@
+                logger.warning(msg)
+                warn(msg)
+
+            if self.lr_data is not None and self.val_split > 0.0:
+                train_indices = list(set(train_indices // self.t_enhance))
+                val_indices = list(set(val_indices // self.t_enhance))
+
+                self.lr_val_data = self.lr_data[:, :, val_indices, :]
+                self.lr_data = self.lr_data[:, :, train_indices, :]
+
+                self.lr_val_time_index = self.lr_time_index[val_indices]
+                self.lr_time_index = self.lr_time_index[train_indices]
+
+                msg = (
+                    'Low res validation data has shape='
+                    f'{self.lr_val_data.shape} and sample_shape='
+                    f'{self.lr_sample_shape}. Use a smaller sample_shape '
+                    'and/or larger val_split.'
+                )
+                check = any(
+                    val_size < samp_size
+                    for val_size, samp_size in zip(
+                        self.lr_val_data.shape, self.lr_sample_shape
+                    )
+                )
+                if check:
+                    logger.warning(msg)
+                    warn(msg)
+
+    def normalize(self, means, stdevs):
+        """Normalize low_res data
+
+        Parameters
+        ----------
+        means : np.ndarray
+            dimensions (features)
+            array of means for all features with same ordering as data features
+        stdevs : np.ndarray
+            dimensions (features)
+            array of means for all features with same ordering as data features
+        """
+        self._normalize(
+            data=self.lr_data,
+            val_data=self.lr_val_data,
+            means=means,
+            stds=stdevs,
+            max_workers=self.lr_dh.norm_workers,
+        )
+
+    @property
+    def output_features(self):
+        """Get list of output features. e.g. those that are returned by a
+        GAN"""
+        return self.hr_dh.output_features
+
+    def _shape_check(self):
+        """Check if hr_handler.shape is divisible by s_enhance. If not take
+        the largest shape that can be."""
+
+        if self.hr_data is None:
+            logger.info("Loading high resolution cache.")
+            self.hr_dh.load_cached_data(with_split=False)
+
+        msg = (
+            'hr_handler.shape is not divisible by s_enhance. Using '
+            f'shape = {self.hr_required_shape} instead.'
+        )
+        if self.hr_dh.shape[:-1] != self.hr_required_shape:
+            logger.warning(msg)
+            warn(msg)
+
+        self.hr_data = self.hr_dh.data[
+            : self.hr_required_shape[0],
+            : self.hr_required_shape[1],
+            : self.hr_required_shape[2],
+        ]
+        self.hr_time_index = self.hr_dh.time_index[: self.hr_required_shape[2]]
+        self.lr_time_index = self.lr_dh.time_index[: self.lr_required_shape[2]]
+
+    def _run_pair_checks(self, hr_handler, lr_handler):
```

Comment:
maybe this is done somewhere else, but i'd recommend adding a check that the time index attributes match exactly.

Reply:
They shouldnt match exactly since this still allows for temporal enhancement but the shapes are checked. I could add a check that `hr_time_index[::t_enhance] = lr_time_index` though.

Reply:
I think that's worthwhile. The risk here is that for whatever reason the hr/lr datasets aren't of the same weather year. I suppose this might also be an issue for lat/lon although that might be more difficult to check.